// gatsby-source-prismic-graphql/gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _react = _interopRequireDefault(require("react"));

exports.onRenderBody = function (_ref, pluginOptions) {
  var setHeadComponents = _ref.setHeadComponents;
  var repositoryName = pluginOptions.repositoryName,
      accessToken = pluginOptions.accessToken,
      _pluginOptions$previe = pluginOptions.previews,
      previews = _pluginOptions$previe === void 0 ? false : _pluginOptions$previe;

  if (previews === false) {
    // Remove accessToken if previews are disabled
    accessToken = null;
  }

  setHeadComponents([_react.default.createElement("script", {
    key: "plugin-source-prismic-graphql",
    dangerouslySetInnerHTML: {
      __html: "window.___sourcePrismicGraphql = ".concat(JSON.stringify({
        repositoryName: repositoryName,
        accessToken: accessToken,
        previews: previews
      }), "; ")
    }
  })]);

  if (previews) {
    setHeadComponents([_react.default.createElement("script", {
      key: "prismic-config",
      dangerouslySetInnerHTML: {
        __html: "\n            window.prismic = {\n              endpoint: 'https://".concat(repositoryName, ".prismic.io/api/v2'\n            };\n          ")
      }
    }), _react.default.createElement("script", {
      key: "prismic-script",
      type: "text/javascript",
      src: "//static.cdn.prismic.io/prismic.min.js"
    })]);
  }
};
