// gatsby-source-prismic-graphql/gatsby-node.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _path = _interopRequireDefault(require("path"));

var _gatsbyNode = require("gatsby-source-graphql-universal/gatsby-node");

var _utils = require("./utils");

exports.onCreateWebpackConfig = _gatsbyNode.onCreateWebpackConfig;

exports.sourceNodes = function (ref, options) {
  options.fieldName = _utils.fieldName;
  options.typeName = _utils.typeName;

  options.createLink = function () {
    return (0, _utils.PrismicLink)({
      uri: "https://".concat(options.repositoryName, ".prismic.io/graphql"),
      credentials: 'same-origin',
      accessToken: options.accessToken
    });
  };

  return (0, _gatsbyNode.sourceNodes)(ref, options);
};

exports.createPages = function (_ref, options) {
  var actions = _ref.actions;
  var previewPath = options.path || '/preview';
  actions.createPage({
    path: previewPath.replace(/^\//, ''),
    component: _path.default.resolve(_path.default.join(__dirname, 'PreviewPage.js')),
    context: {
      repositoryName: options.repositoryName
    }
  });
};
