// gatsby-source-prismic-graphql/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "withPreview", {
  enumerable: true,
  get: function get() {
    return _withPreview.withPreview;
  }
});
Object.defineProperty(exports, "PreviewPage", {
  enumerable: true,
  get: function get() {
    return _PreviewPage.PreviewPage;
  }
});
Object.defineProperty(exports, "PrismicLink", {
  enumerable: true,
  get: function get() {
    return _utils.PrismicLink;
  }
});
Object.defineProperty(exports, "getCookies", {
  enumerable: true,
  get: function get() {
    return _utils.getCookies;
  }
});
Object.defineProperty(exports, "registerResolvers", {
  enumerable: true,
  get: function get() {
    return _utils.registerResolvers;
  }
});
Object.defineProperty(exports, "linkResolver", {
  enumerable: true,
  get: function get() {
    return _utils.linkResolver;
  }
});
Object.defineProperty(exports, "componentResolver", {
  enumerable: true,
  get: function get() {
    return _utils.componentResolver;
  }
});
Object.defineProperty(exports, "fieldName", {
  enumerable: true,
  get: function get() {
    return _utils.fieldName;
  }
});
Object.defineProperty(exports, "typeName", {
  enumerable: true,
  get: function get() {
    return _utils.typeName;
  }
});

var _withPreview = require("./withPreview");

var _PreviewPage = require("./PreviewPage");

var _utils = require("./utils");
