// gatsby-source-prismic-graphql/PreviewPage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PreviewPage = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _prismicJavascript = _interopRequireDefault(require("prismic-javascript"));

var _utils = require("./utils");

var _withPreview = require("./withPreview");

;

var PreviewPage =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(PreviewPage, _React$Component);

  function PreviewPage() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, PreviewPage);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(PreviewPage)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "url", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "qs", new Map());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "state", {
      component: null
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)), "redirect",
    /*#__PURE__*/
    function () {
      var _ref = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee(doc) {
        var link, exists;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (doc) {
                  _context.next = 3;
                  break;
                }

                window.location = '/';
                return _context.abrupt("return");

              case 3:
                link = (0, _utils.linkResolver)(doc);
                _context.next = 6;
                return fetch(link).then(function (res) {
                  return res.status;
                });

              case 6:
                _context.t0 = _context.sent;
                exists = _context.t0 === 200;

                if (!exists) {
                  _context.next = 11;
                  break;
                }

                window.location = link;
                return _context.abrupt("return");

              case 11:
                if (typeof _utils.componentResolver === 'function') {
                  _this.setState({
                    component: (0, _utils.componentResolver)(doc)
                  });

                  window.history.replaceState({}, 'prismic', window.location.pathname + '?documentId=' + doc.id);
                }

              case 12:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function (_x) {
        return _ref.apply(this, arguments);
      };
    }());
    return _this;
  }

  (0, _createClass2.default)(PreviewPage, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.setup();
    }
  }, {
    key: "setup",
    value: function setup() {
      if (typeof window !== 'undefined') {
        this.url = new URL(window.location.toString());
        this.qs = (0, _utils.qs)(String(this.url.search).substr(1));
        this.preview();
      }
    }
  }, {
    key: "preview",
    value: function () {
      var _preview = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee2() {
        var token, experiment, documentId, now, api, doc, runningVariations, matchedVariation, cookies, _doc, preview;

        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                token = this.qs.get('token');
                experiment = this.qs.get('experiment');
                documentId = this.qs.get('documentId'); // Expiration date of cookie

                now = new Date();
                now.setHours(now.getHours() + 1);
                _context2.next = 7;
                return _prismicJavascript.default.getApi("https://".concat(this.config.repositoryName, ".cdn.prismic.io/api/v2"));

              case 7:
                api = _context2.sent;

                if (!token) {
                  _context2.next = 20;
                  break;
                }

                _context2.next = 11;
                return api.previewSession(token, _utils.linkResolver, '/');

              case 11:
                document.cookie = "".concat(_prismicJavascript.default.previewCookie, "=").concat(token, "; expires=").concat(now.toUTCString(), "; path=/");

                if (documentId) {
                  _context2.next = 14;
                  break;
                }

                return _context2.abrupt("return", this.redirect());

              case 14:
                _context2.next = 16;
                return api.getByID(documentId);

              case 16:
                doc = _context2.sent;
                return _context2.abrupt("return", this.redirect(doc));

              case 20:
                if (!experiment) {
                  _context2.next = 26;
                  break;
                }

                runningVariations = [];

                if (api.experiments.running && api.experiments.running.length) {
                  runningVariations.concat.apply(runningVariations, (0, _toConsumableArray2.default)(api.experiments.running.map(function (experiment) {
                    return experiment.data.variations;
                  })));
                }

                if (experiment && runningVariations.length) {
                  matchedVariation = runningVariations.find(function (variation) {
                    return variation.label.toLowerCase().replace(' ', '-') === experiment;
                  });

                  if (matchedVariation) {
                    document.cookie = "".concat(_prismicJavascript.default.experimentCookie, "=").concat(matchedVariation.ref, "; expires=").concat(now.toUTCString(), "; path=/");
                    this.redirect();
                  }
                }

                _context2.next = 33;
                break;

              case 26:
                if (!documentId) {
                  _context2.next = 33;
                  break;
                }

                cookies = (0, _utils.getCookies)();
                _context2.next = 30;
                return api.getByID(documentId);

              case 30:
                _doc = _context2.sent;
                preview = cookies.has(_prismicJavascript.default.previewCookie) || cookies.has(_prismicJavascript.default.experimentCookie);
                this.redirect(preview && _doc);

              case 33:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      return function preview() {
        return _preview.apply(this, arguments);
      };
    }()
  }, {
    key: "render",
    value: function render() {
      if (this.state.component) {
        var component = this.state.component.default || this.state.component;

        if (component.query) {
          return _react.default.createElement((0, _withPreview.withPreview)(component, component.query));
        }

        return _react.default.createElement(component);
      }

      return null;
    }
  }, {
    key: "config",
    get: function get() {
      if (typeof window !== 'undefined') {
        var config = window.___sourcePrismicGraphql;
        return config || {};
      }

      return {};
    }
  }]);
  return PreviewPage;
}(_react.default.Component);

exports.PreviewPage = PreviewPage;
var _default = PreviewPage;
exports.default = _default;
